src/TopDownProteomics/ProForma/TopPicProformaParser.cs

```diff
@@ -0,0 +1,186 @@
+    {
+        IDictionary<string, ProFormaDescriptor>? _modLookup = null;
+
+        #region Regex strings
+        Regex _modRx = new Regex(@"\(([A-Z]{1,})\)(\[[\w_+-.]+\])+");
+        Regex _numberRx = new Regex(@"(-?\+?[0-9]+.[0-9]+)");
+        Regex _terminalAaRx = new Regex(@"\P{N}(\.)\P{N}??|\P{N}??(\.)\P{N}");
+        #endregion
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TopPicProformaParser"/> class.
+        /// </summary>
+        public TopPicProformaParser() { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TopPicProformaParser"/> class.
+        /// </summary>
+        /// <param name="modFile">The mod.txt file for mapping modifications.</param>
+        public TopPicProformaParser(string modFile)
+        {
+            _modLookup = ParseModFile(modFile);
+        }
+
+        /// <summary>
+        /// Gets the proforma term.
+        /// </summary>
+        /// <param name="sequence">The sequence.</param>
+        /// <returns></returns>
+        public ProFormaTerm ParseTopPicString(string sequence)
+        {
+            //first remove terminaltags if there!
+            sequence = RemoveTerminalAAs(sequence);
+            var ptms = FindPTMs(sequence);
+            return new ProFormaTerm(GetFullyStrippedSequence(sequence), ptms.Item3, ptms.Item1, ptms.Item2);
+        }
+
+        private IDictionary<string, ProFormaDescriptor> ParseModFile(string modFile)
+        {
+            IDictionary<string, ProFormaDescriptor> modLookup = new Dictionary<string, ProFormaDescriptor>();
+
+            using StreamReader reader = new StreamReader(modFile);
+
+            while (!reader.EndOfStream)
+            {
+                var line = reader.ReadLine();
+
+                if (line.Length == 0 | line.StartsWith("#"))
+                    continue;
+
+                //# To input a modification, use the following format:
+                //# Name,Mass,Residues,Position,UnimodID
+
+                var splitLine = line.Split(',');
+                var name = splitLine[0];
+
+                if (Int32.TryParse(splitLine[4], out var uniModNumber))
+                {
+                    if (uniModNumber > 0)
+                        modLookup.Add(name, new ProFormaDescriptor(ProFormaKey.Identifier, ProFormaEvidenceType.Unimod, $"UNIMOD:{uniModNumber}"));
+                    else if (uniModNumber == -1)
+                        modLookup.Add(name, new ProFormaDescriptor(ProFormaKey.None, name)); // maybe could do mass instead?
+                    else
+                        throw new Exception($"invalid unimod id");
+
+                }
+                else
+                    throw new Exception($"Failed to parse unimod id {splitLine[1]}");
+            }
+
+            return modLookup;
+        }
+
+
+        private string GetFullyStrippedSequence(string sequence) => Regex.Replace(sequence, @"\[[\w_+-.]+\]|[()]", "");
```

Comment:
what is the first clause in the regex trying to do? It seems like there are missing characters that might cause problems ...

Reply:
it needs to pull out everything inside []. I tried a couple of things. but you can't have something that is inclusive of ']' or it breaks!

Reply:
`\[[^]^[]+\]|[()]` 
This seems to work